WooCommerce/Classes/ViewRelated/Products/Edit Product/Product Settings/Catalog Visibility/ProductCatalogVisibilityViewController.swift

```diff
@@ -0,0 +1,234 @@
+//
+extension ProductCatalogVisibilityViewController: UITableViewDelegate {
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        tableView.deselectRow(at: indexPath, animated: true)
+        let row = sections[indexPath.section].rows[indexPath.row]
+        if let catalogVisibility = row.catalogVisibility {
+            productSettings.catalogVisibility = catalogVisibility
+        }
+        tableView.reloadData()
+    }
+
+    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
+        return UITableView.automaticDimension
+    }
+
+    func tableView(_ tableView: UITableView, estimatedHeightForHeaderInSection section: Int) -> CGFloat {
+        return CGFloat.leastNormalMagnitude
+    }
+
+    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
+
+        let headerID = TwoColumnSectionHeaderView.reuseIdentifier
+        guard let headerView = tableView.dequeueReusableHeaderFooterView(withIdentifier: headerID) as? TwoColumnSectionHeaderView else {
+            assertionFailure("Unregistered \(TwoColumnSectionHeaderView.self) in UITableView")
+            return nil
+        }
+
+        headerView.leftText = nil
+        headerView.rightText = nil
+
+        return headerView
+    }
+}
+
+// MARK: - Support for UITableViewDataSource
+//
+private extension ProductCatalogVisibilityViewController {
+    /// Configure cellForRowAtIndexPath:
+    ///
+   func configure(_ cell: UITableViewCell, for row: Row, at indexPath: IndexPath) {
+        switch cell {
+        case let cell as SwitchTableViewCell:
+            configureFeaturedProduct(cell: cell)
+        case let cell as BasicTableViewCell:
+            configureCatalogVisibilitySelector(cell: cell, indexPath: indexPath)
+        default:
+            fatalError("Unidentified product catalog visibility row type")
+        }
+    }
+
+    func configureFeaturedProduct(cell: SwitchTableViewCell) {
+        cell.title = NSLocalizedString("Featured Product", comment: "Featured Product switch in Product Catalog Visibility")
+        cell.isOn = productSettings.featured
+        cell.onChange = { [weak self] value in
+            self?.productSettings.featured = value
+        }
+    }
+
+    func configureCatalogVisibilitySelector(cell: BasicTableViewCell, indexPath: IndexPath) {
+        let row = sections[indexPath.section].rows[indexPath.row]
+        cell.selectionStyle = .default
+        cell.textLabel?.text = row.catalogVisibility?.description
+        cell.accessoryType = productSettings.catalogVisibility == row.catalogVisibility ? .checkmark : .none
+    }
+}
+
+// MARK: - Constants
+//
+extension ProductCatalogVisibilityViewController {
+
+    /// Table Rows
+    ///
+    enum Row {
+        /// Listed in the order they appear on screen
+        case featuredProduct
+        case visible
+        case catalog
+        case search
+        case hidden
```

Comment:
nit: the catalog visibility cases sound a bit generic and could be used in other context, would adding a prefix like `catalogVisibility*` (or some other naming) be a bit more clear?

another idea: how about a case with associated value like `case catalogVisibility(ProductCatalogVisibility)`?